src/aead.rs

```diff
@@ -236,6 +267,49 @@ fn open_within_<'in_out, A: AsRef<[u8]>>(
     )
 }
 
+#[inline]
+fn open_in_place_separate_tag<'in_out, A: AsRef<[u8]>>(
+    key: &UnboundKey,
+    nonce: Nonce,
+    Aad(aad): Aad<A>,
+    in_out: &'in_out mut [u8],
+    received_tag: &[u8],
+) -> Result<&'in_out mut [u8], error::Unspecified> {
+    fn open_within<'in_out>(
+        key: &UnboundKey,
+        nonce: Nonce,
+        aad: Aad<&[u8]>,
+        in_out: &'in_out mut [u8],
+        received_tag: &[u8; TAG_LEN],
+    ) -> Result<&'in_out mut [u8], error::Unspecified> {
+        let ciphertext_len = in_out.len();
+        check_per_nonce_max_bytes(key.algorithm, ciphertext_len)?;
+        let Tag(calculated_tag) =
+            (key.algorithm.open)(&key.inner, nonce, aad, 0, in_out, key.cpu_features);
+        if constant_time::verify_slices_are_equal(calculated_tag.as_ref(), received_tag).is_err() {
+            // Zero out the plaintext so that it isn't accidentally leaked or used
+            // after verification fails. It would be safest if we could check the
+            // tag before decrypting, but some `open` implementations interleave
+            // authentication with decryption for performance.
+            for b in in_out {
+                *b = 0;
+            }
+            return Err(error::Unspecified);
+        }
+        // Ciphertext is now decrypted to plaintext.
+        Ok(in_out)
+    }
+
+    use core::convert::TryInto;
+    open_within(
+        key,
+        nonce,
+        Aad::from(aad.as_ref()),
+        in_out,
+        received_tag.try_into()?,
+    )
```

Comment:
This duplicates some functionality that already exists in `open_within_()`. Maybe it would be better to change `open_within_()` to accept `ciphertext` and `tag` separately? Or have `open_within_()` depend on an `open_within_separate_tag_()` function?

Reply:
Personally I'm fine with the amount of duplication here. `open_in_place_separate_tag_` is really simple and straightforward. A `open_within_separate_tag_` would be very complex, I don't think it's worth it.